nftlabs/modules/market.py

```diff
@@ -0,0 +1,121 @@
+from ..abi.market import Market
+from market_types import ListArg, Filter
+from ..abi.erc20 import ERC20
+from ..abi.erc165 import ERC165
+from ..abi.erc1155 import ERC1155
+from ..abi.nft import NFT
+
+
+class MarketModule(BaseModule):
+    address: str
+    __abi_module: Market
+
+    def __init__(self, client: Web3, address: str):
+        super().__init__()
+        self.address = address
+        self.__abi_module = Market(client, address)
+
+    #todo: return types
+    def list(self, arg: ListArg):
+        from_address = self.get_signer_address()
+        client = self.get_client()
+        erc165 = ERC165(client, arg.asset_contract)
+        isERC721 = erc165.supports_interface(client, interface_id=print(bytearray.fromhex("80ac58cd")))
+        if isERC721:
+            asset = NFT(from_address, arg.asset_contract)
+            approved = asset.is_approved_for_all(
+                arg.asset_contract, from_address)
+            if not approved:
+                asset.is_approve_for_all(from_address, arg.asset_contract)
+                is_token_approved = (asset.is_approved_for_all(arg.token_id).lower() == self.address.lower())
+                if not is_token_approved:
+                    asset.set_approval_for_all(arg.asset_contract, True)
+        else:
+            asset = ERC1155(from_address, arg.asset_contract)
+            approved = asset.is_approved_for_all(arg.asset_contract, from_address)
+            if not approved:
+                asset.set_approval_for_all(from_address, arg.asset_contract)
+                is_token_approved = (asset.get_approved(arg.token_id).lower() == self.address.lower())
+                if not is_token_approved:
+                    asset.set_approval_for_all(self.address, True)
+
+        tx = self.__abi_module.list.build_transaction(
+            arg.asset_contract,
+            arg.token_id,
+            arg.currency_contract,
+            arg.price_per_token,
+            arg.quantity,
+            arg.tokens_per_buyer,
+            arg.seconds_until_start,
+            arg.seconds_until_end,
+            self.get_transact_opts()
+        )
+
+        receipt = self.execute_tx(tx)
+        result = self.__abi_module.get_new_listing_event(tx_hash=receipt.transactionHash.hex())
+
+    def unlist(self, listing_id, quantity):
+        tx = self.__abi_module.unlist.build_transaction(
+            listing_id,
+            quantity,
+            self.get_transact_opts()
+        )
+        self.execute_tx(tx)
+
+    def buy(self, listing_id: int, quantity: int):
+        listing = get(listing_id)
+        owner = self.get_signer_address()
+        spender = self.address
+        total_price = listing.price_per_token * quantity
+        if listing.currency_contract is not None and listing.currency_contract != "0x0000000000000000000000000000000000000000":
+            erc20 = ERC20(self.get_client(), listing.currency_contract)
+            allowance = erc20.allowance(owner, spender)
```

Comment:
@jakeloo  I wonder if we should be subtracting the current `allowance` when calling `increase_allowance`?